core/templates/services/math-interactions.service.ts

```diff
@@ -61,9 +61,27 @@ export class MathInteractionsService {
         'not valid.';
     }
     if (errorMessage.indexOf('is not a valid postfix operator.') !== -1) {
+      let humanReadableOperator = (
+        errorMessage[0] === '*' ? '×' :
+        errorMessage[0] === '/' ? '÷' :
+        errorMessage[0]);
       errorMessage = (
-        'Your answer seems to be missing a variable/number after the "' +
-        errorMessage[0] + '".');
+        'Your answer seems to be missing a variable/number after the ' +
+        humanReadableOperator + ' operator.');
+    }
+    if (errorMessage === 'Not a prefix operator.') {
+      let i: number;
+      for (i = 0; i < expressionString.length; i++) {
+        if (expressionString[i] === '*' || expressionString[i] === '/') {
+          let humanReadableOperator = (
+            expressionString[i] === '*' ? '×' :
+            expressionString[i] === '/' ? '÷' :
+            null);
```

Comment:
Have you tested this with more complex expressions, such as "*3 / 4" or "4 / (* 2) / 3? What is your full reasoning that this code is correct?

Reply:
Actually in the nerdamer code [here](https://github.com/jiggzson/nerdamer/blob/28d5ca348ad881b765615e56dacfb82521050ee2/nerdamer.core.js#L6510-L6529) it is seen in line 6265 that nerdamer message would be `Not a prefix operator` when there is no any operand before the operator.
So I have written this code and think it is correct because of following reasons: 
1. if operators are either `* and /` then my code will update the error message and make it more readable. like 
`Your answer seems to be missing a variable/number after the × operator.`-> for operator `*` and 
`Your answer seems to be missing a variable/number after the ÷ operator.`-> for operator `/`
2. If there is any other operator which gives nerdamer error as `Not a prefix operator.` then the updated error message would be `Your answer seems to be missing a variable/number after the operator.`

And for your other question-> the answer is yes, I have checked and it is working fine for more complex expressions like above. Here is the proof.

[Screencast from 23-02-23 10:59:26 PM IST.webm](https://user-images.githubusercontent.com/91023072/220986994-554779d1-cd01-443a-be64-054f18f68bcd.webm)

Reply:
@adityanarayanm095 I don't understand what your video proves, since you are entering only valid expressions as far as I can tell. Could you please try the expressions I gave you instead? (Also, sorry, I just realized there was an issue in my original comment where github interpreted * as italics -- I've fixed this.

Reply:
Hey, I am working on it.

Reply:
@seanlip 
I have tried with different codes to update the error message but It is not updating for complex expressions and if it fixes for one kind but not for another kind of complex expression: 
I tried-> 
```
    if (errorMessage === 'Not a prefix operator.') {
      let i: number;
      for (i = 0; i < expressionString.length; i++) {
        if ((expressionString.charAt(i) === '*') &&
        (typeof expressionString[i - 1] !== 'number' &&
        typeof expressionString[i - 1] !== 'string') ||
       (typeof expressionString[i + 1] !== 'number' &&
        typeof expressionString[i + 1] !== 'string')) {
          errorMessage = (
            'Your answer seems to be missing a variable/number before or after the × operator'
          );
        } else if ((expressionString.charAt(i) === '/') &&
        (typeof expressionString[i - 1] !== 'number' &&
        typeof expressionString[i - 1] !== 'string') ||
       (typeof expressionString[i + 1] !== 'number' &&
        typeof expressionString[i + 1] !== 'string')) {
          errorMessage = (
            'Your answer seems to be missing a variable/number before or after the ÷ operator'
          );
        }
      }
    }
```

    In that case can we simply go with the generalised error message which will work for all the operators. eg-> 

        if (errorMessage === 'Not a prefix operator.') {
      errorMessage = 'Your answer seems to be missing a variable/number' +
        ' before the operator.';
    }
```

But if you any other code suggestion, please let me know!

Reply:
@seanlip Any update?

Reply:
A string matching doesn't really work for this sort of validation. Is it possible to get any other info at all from nerdamer besides the error string? What other info is provided to us / can be provided to us by the nerdamer functionality?

Reply:
Hey @seanlip 
1. I found out that, Nerdamer operator is defined as custom error in the file which you can check here-> https://github.com/jiggzson/nerdamer/blob/28d5ca348ad881b765615e56dacfb82521050ee2/nerdamer.core.js#L1279
2. Here is the definition of custom error-> https://github.com/jiggzson/nerdamer/blob/28d5ca348ad881b765615e56dacfb82521050ee2/nerdamer.core.js#L187-L197
3. By analysis using `console.error(err)` in our codebase 
![image](https://user-images.githubusercontent.com/91023072/224383440-0dcc130a-a620-44e6-b96f-25a1440aef31.png)
we are getting error message in console as-> 
![image](https://user-images.githubusercontent.com/91023072/224384348-2a328860-b7f8-4515-8d38-29b002fe9b5b.png)

```{
    "name": "OperatorError",
    "message": "Not a prefix operator",
    "stack": "OperatorError: Not a prefix operator\n    at new E (webpack:///./node_modules/nerdamer/nerdamer.core.js?:191:25)\n    at Parser.toRPN (webpack:///./node_modules/nerdamer/nerdamer.core.js?:6526:27)\n    at Parser.parse (webpack:///./node_modules/nerdamer/nerdamer.core.js?:6944:28)\n    at eval (webpack:///./node_modules/nerdamer/nerdamer.core.js?:12042:22)\n    at block (webpack:///./node_modules/nerdamer/nerdamer.core.js?:1094:24)\n    at libExports (webpack:///./node_modules/nerdamer/nerdamer.core.js?:12041:17)\n    at MathInteractionsService.insertMultiplicationSigns (webpack:///./core/templates/services/math-interactions.service.ts?:371:72)\n    at MathInteractionsService._validateExpression (webpack:///./core/templates/services/math-interactions.service.ts?:206:37)\n    at MathInteractionsService.validateNumericExpression (webpack:///./core/templates/services/math-interactions.service.ts?:273:19)\n    at InteractiveNumericExpressionInput.isCurrentAnswerValid (webpack:///./extensions/interactions/NumericExpressionInput/directives/oppia-interactive-numeric-expression-input.component.ts?:77:63)"
}

Reply:
So @seanlip either we have to change error message from nerdamer itself?
Please tell what should we do then in this case?

Reply:
Good find. Alas, this means that you probably can't get any further info from the nerdamer error. (I was hoping it would have some kind of parser breakdown or a more explicit location in the string at which the parser failed.)

I think then you'll need to adjust the error message so that it's never wrong, by scanning the string. Check the nerdamer code: is it only × and ÷ that would result in this error message, or can + or - do so as well? Based on this, you can format your message accordingly -- scan the input string and look for what "problematic" operators there might be in it. If there's just one, you can write a definite error message. If not then you can write 

            'Your answer seems to be missing a variable/number before one of the operators ({{insert your list of problematic operators here}})'.

and that would be a true statement, although slightly les specific.

Reply:
@seanlip As you can see all the operators defined [here](https://github.com/jiggzson/nerdamer/blob/28d5ca348ad881b765615e56dacfb82521050ee2/nerdamer.core.js#L5547) and in the input filed I can see that it is accepting the answer for `+ and -` as per expected and giving problem only for `multiplication and division sign` as you can check in the video 
[Screencast from 13-03-23 05:10:21 AM IST.webm](https://user-images.githubusercontent.com/91023072/224581178-ff0db7f1-6fd0-4ec0-a95c-904318891f94.webm)

My concern is if user gives answer firstly as-> `3*4/2(*2)/4` and our error messages comes `Your answer seems to be missing a variable/number before one of the operators (×, ÷)` then he may give answer `3*4/2(*2)2/4` again it will give the same error message, then finally he will give answer `3*4/2(2*2)2/4` now this time the error message `Your answer seems to be missing a variable/number before one of the operators (×, ÷)` would not come but a feedback message will come
[Screencast from 13-03-23 05:18:41 AM IST.webm](https://user-images.githubusercontent.com/91023072/224581567-36d2d02a-283e-4c18-a01d-6505e0ddf7de.webm)
so, the learner may get confused. What can be done in that case?

Reply:
Hi @adityanarayanm095, I'm a bit confused. That latter sequence seems fine, I think? The user would need to click Submit before they get the feedback message anyway. Or am I missing something here?

Also just to check, for `2(*2)` the error message you have only references × and not ÷, right?

Reply:
@seanlip The  easiest errorMessage to cover all the above conflicts is `Your answer seems to be missing a variable/number before one of the operator(s).`  so that user will not get confused and easily find out the operator before which a number is missing. This is only I am able to think at this point of time.
otherwise it is quite difficult to cover all the cases, there may be so many cases like above you are mentioning. Should I introduce it?

Other option is-> Can you please suggest me the code to cover all the possible cases o/w?

Reply:
We should try to give the most specific error message possible, but it has to be correct. How about this:

If it's numeric, then:
- If there's only one possible operator, e.g. no ÷ appears in the expression: `Your answer seems to be missing a number before the operator ×` (or ÷ depending on context)
- If there can be multiple possibilities: `Your answer seems to be missing a number before one of the operators (×, ÷)`

If it's algebraic, then replace "number" above with "variable/number".

Please check this for correctness. Am I missing anything here?